### PR TITLE
Amr Yehia week4-JavaScript

### PR DESCRIPTION
**Pull request by Amr-Yehia**

### CODE REVIEW

**Comment by robvk** [comment]
Hi @Amr-Yehia , 
You have also put week3's files into your PR which is confusing. For this time we will let it go, but when creating your branch for a new week make sure to create it from the `main` branch and not the previous week's branch.

**Comment by Amr-Yehia** [comment]
Hi @robvk 

Oh sorry,I will keep that in mind for next weeks.
Thank you :)

File: 1-JavaScript/Week4/homework/ex1-doubleEvenNumbers.test.js [hunk 1]
```diff
@@ -1,29 +1,26 @@
...
-    }
-  }
-  return newNumbers;
+  const newNumbers = numbers.filter((num) => {
+    return num %2 === 0;
+  });
+  const doubleNum = newNumbers.map((num) => {
+    return num * 2;
+  });
+  return doubleNum;
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week4/homework/ex1-doubleEvenNumbers.test.js:18]: Great job 👏 You could consider using the shorter syntax of the arrow function:
```javascript
numbers.filter(num => num % 2 === 0)
```
Whatever you find more readable 🙂

File: 1-JavaScript/Week4/homework/ex2-mondaysWorth.test.js [hunk 2]
```diff
@@ -31,8 +31,15 @@ const mondayTasks = [
...
-function computeEarnings(/* TODO parameter(s) go here */) {
+function computeEarnings( mondayTasks, hourlyRate/* TODO parameter(s) go here */) {
   // TODO complete this function
+  const durations = mondayTasks.map(time => {
+    return time.duration/60 * hourlyRate;
+  });
+  const sum = durations.reduce((actualNumber, calculateNumber) => {
+    return actualNumber + calculateNumber;
+  });
+  return `€${sum.toFixed(2)}`;
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week4/homework/ex2-mondaysWorth.test.js:42]: Perfect! Something to consider are the names of your parameters of the reducer function (`actualNumber` and `calculateNumber`). They might not be too distinguishable. Sometimes it is hard to think of something and then you can always fallback to `acc` and `cur`.

Also you can chain the map and reduce so you don't have to store it in consts, but that is just detail. Overall you implemented a great solution.

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js [hunk 3]
```diff
@@ -26,26 +26,34 @@ const fruitBasket = [
 ];
 
 // ! Function under test
-function sanitizeFruitBasket(/* TODO parameter(s) go here */) {
+function sanitizeFruitBasket(basket, item/* TODO parameter(s) go here */) {
   // TODO complete this function
+  const newBasket = basket.filter(fruit => fruit !== item);
+  
+  return newBasket;
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:33]: Your solution works like a charm. You might want to consider returning the basket.filter(...) directly

File: 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js [hunk 4]
```diff
@@ -26,26 +26,34 @@ const fruitBasket = [
...
-    expect(false).toBe(true);
+    expect(sanitizeFruitBasket).toHaveLength(2);
   });
 
   test('should not modify the original `fruitBasket` array', () => {
     // Save the original contents of the fruit basket
     const originalFruitBasketContents = [...fruitBasket];
     // TODO replace next line with your code
-    expect(false).toBe(true);
+    expect(originalFruitBasketContents).toEqual(fruitBasket);
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week4/homework/ex3-lemonAllergy.test.js:47]: Your test passes, but you did not call the sanitize function before the expect. The idea is to test if `fruitbasket` stays the same if you input it to the sanitise function.

File: 1-JavaScript/Week4/homework/ex4-observable/ex4-observable.js [hunk 5]
```diff
@@ -18,9 +18,12 @@ function createObservable() {
   return {
     subscribe: function (subscriber) {
       // TODO complete this function
+      subscribers.push(subscriber);
+
     },
     notify: function (message) {
       // TODO complete this function
+      subscribers.forEach((item) => item(message));
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week4/homework/ex4-observable/ex4-observable.js:26]: well done!

**Review by bnhoogendijk — APPROVED** [review]
Excellent work this week 🎉 I left you some comments but no need for extra commits.

**Pull request closed without merging**